### PR TITLE
[tune] Fix Pausing and Error Propogation

### PR DESCRIPTION
**Pull request by richardliaw**
## What do these changes do?
Addresses #2814 and #2810.

 - Manages error handling with recovery, even when error occurs at `ray.get`.
 - Also introduces tests for pausing via trial executor.

## Related issue number
Above.

### CODE REVIEW

**Comment by ericl** [comment]
Can you also change the smoke test iterations for the pbt test to something like 20? That catches the bug and also runs reasonably fast.

**Comment by richardliaw** [comment]
@joyyoj Can you take a look please?

**Comment by richardliaw** [comment]
jenkins retest this please

**Comment by richardliaw** [comment]
Jenkins retest this please

File: python/ray/tune/trial_executor.py [hunk 1]
```diff
@@ -119,17 +119,23 @@ def on_step_end(self):
...
+        Assumes the trial is running.
 
         Return:
-            A tuple of (trial, result). If fetch result failed,
-            return (trial, None) other than raise Exception.
+            Result object for the trial.
         """
-        raise NotImplementedError("Subclasses of TrialExecutor must provide "
-                                  "fetch_one_result() method")
+        raise NotImplementedError
```
Comment (ericl) [inline python/ray/tune/trial_executor.py]: So one other way of doing this is to have fetch_one_result() return an exception object for the result if there is an error. This avoids having a separate call just for this error handling.

**Review by ericl — APPROVED** [review]
lgtm based on tests and @joyyoj 's approval (I did not look carefully at the unpause crash, it looks complicated)

One minor suggestion on the error handling that might give a cleaner interface.

**Pull request merged**